build/parsePreamble.c

```diff
@@ -1267,28 +1267,31 @@ int parsePreamble(rpmSpec spec, int initialPackage)
 	}
     }
 
-    /* 
-     * Expand buildroot one more time to get %{version} and the like
-     * from the main package, validate sanity. The spec->buildRoot could
-     * still contain unexpanded macros but it cannot be empty or '/', and it
-     * can't be messed with by anything spec does beyond this point.
-     */
     if (initialPackage) {
 	if (checkForRequiredForBuild(pkg->header)) {
 	    goto exit;
 	}
 
-	char *buildRoot = rpmGetPath(spec->buildRoot, NULL);
-	free(spec->buildRoot);
-	spec->buildRoot = buildRoot;
-	rpmPushMacro(spec->macros, "buildroot", NULL, spec->buildRoot, RMIL_SPEC);
-	if (*buildRoot == '\0') {
-	    rpmlog(RPMLOG_ERR, _("%%{buildroot} couldn't be empty\n"));
-	    goto exit;
+	/* Grab the top builddir on first entry as we'll override _builddir */
+	if (!rpmMacroIsDefined(spec->macros, "_top_builddir")) {
+	    rpmPushMacro(spec->macros, "_top_builddir", NULL, "%{_builddir}",
```

Comment:
I may be confused, but don't you need to expand %{_buildir} first?

Reply:
This seems wrong, indeed; the value is (recursively) expanded through `%{_top_builddir}` in the `rpmExpand()` call right away, *but* that call only happens if `spec->buildDir` is unset.

Reply:
Oh indeed. The logic for setting %builddir does work, but %_top_builddir will end up pointing to %builddir instead which is not the intention.

I'll fixup and make all these macros use RPMMACRO_LITERAL, that's the right thing here really.